### PR TITLE
Feat: Grand package refactor

### PR DESCRIPTION
**Pull request by ahmedheakl**
Feature - Logging and Refactoring

## 📜 Summary
Since most drone users need some post-monitoring abilities to track the flight changes and processes running over the drone, a `logger` module must be included in the library to provide users such features. Moreover, some modules in the library are missing docstrings, don't comply to PEP8 style standards, don't implement the best software engineering practices. Henceforth, 
`mypy` and `pylint` would be used to refactor the code for better styling. In addition, `mypy` and `pylint` would be added to the CI to provide code integrity between contributors. 

- 🔄 **Main changes in this PR**
    - Added logging module for both `CLI` and `GUI` main scripts.
    - Added file/stream handlers for the logging module.
    - Added `mypy` and `pylint` to the github CI.

- 🛠 **Other changes**:
    - Add CI, add mypy, address pylint issues...
    - Upgraded library major version since backward compatibility will be broken after this release.
    - Added more tests for higher code coverage.

- ⚠️ **Breaking changes**
  - All models now are contained in a single module `models` for better integrity.
  - `cli` and `gui` and now moved into a single module `ui`.
  - `utils.utils.py` is renamed to `utils.general.py`.
  - `logger.py` module is merged into `utils.general.py` module.
  - `gui.configs.py` is moved to `dronevis.configs.gui.py`.
  - `configs.configs.py` is renamed to `configs.general.py`.

## 🚀 How to test this Pull Request

   - For tests, run `pytest tests`
   - To initialize logger instance, there are two scenarios you can have: 
      . Running from an implementation file. 
      ```python
      import logging
      _LOG = logging.getLogger(__name__) # Get logger instance
      _LOG.info("<your-logs>") # Output to console
      ```
      . Running from a script file. You need to initialize the logger configs first. 
      ```python
      import logging
      from dronevis.utils.logger import init_logger
      init_logger(debug=False) # Initialize logger configs
      _LOG = logging.getLogger(__name__)
      _LOG.info("<your-logs>") 
      ```

## 📝 Notes for future improvements

- Refactor `dronevis.ui.main_gui` module.
- Refactor `dronevis.detection_gluoncv` module.
- Handle running `detection_gluoncv` module on Windows OS.
- Add more tests to raise the test coverage for more reliability.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,3 +24,41 @@
 
 
 # DroneVis: Full compatible drone library to automate computer vision algorithms on parrot drones.
+
+**DroneVis** is a cutting-edge drone software library that has been specifically designed for use with the AR. Drone 2.0. It has been extensively tested both indoors and outdoors, and offers a wide range of features including adaptability in connecting to the drone, advanced computer vision algorithms, and a user-friendly interface. This makes it easy for users to take full advantage of the drone's capabilities and control it with simple commands.All of the implemented real-time data, inference, and detection achieve a minimum ``fps >= 4.5`` on an Intel core 8 CPU.
```
Comment (FatmaYoussef96) [inline README.md:28]: space needed after This makes it easy for users to take full advantage of the drone's capabilities and control it with simple commands.

File: README.md [hunk 2]
```diff
@@ -23,3 +24,41 @@
...
+- PEP8 compliant (unified code style)
+- Documented functions and classes
+- Tests, high code coverage and type hints
+- Clean code
+- Multiple implementations for the same models
+- Two UI for easier usage (GUI, CLI)
+
+
+
+| Drone Control         | Computer Vision Models| Usage                 |   Implementation      |
```
Comment (FatmaYoussef96) [inline README.md:42]: The table is a bit confusing to me as I am trying to read one row at a time. 
Is the Drone control column separated from the rest of the table.

File: src/dronevis/abstract/abstract_torch_model.py [hunk 3]
```diff
@@ -49,29 +52,39 @@ def predict(
...
+        input_image = image
         with torch.no_grad():
-            image = self.transform_img(image).to(self.device)
-            image = image.unsqueeze(0)  # add a batch dimension
-            outputs = self.net(image)[0]  # get outputs array
-            self.pred_classes = [self.coco_names[i] for i in outputs["labels"].cpu().numpy()]
-            self.pred_scores = outputs["scores"].detach().cpu().numpy()
-            self.pred_bboxes = outputs["boxes"].detach().cpu().numpy()
-            self.boxes = self.pred_bboxes[self.pred_scores >= detection_threshold].astype(np.int32)
+            transformed_image = self.transform_img(input_image).to(self.device)
```
Comment (FatmaYoussef96) [inline src/dronevis/abstract/abstract_torch_model.py:67]: I think it should be image not input_image

File: src/dronevis/detection_torch/yolov5_torch.py [hunk 4]
```diff
@@ -15,18 +17,27 @@ class YOLOv5(CVModel):
...
+
     def __init__(self) -> None:
         """Initialize local path"""
         self.net = None
-        self.model_local_path = f"/home/{getpass.getuser()}/.cache/torch/hub/ultralytics_yolov5_master"
 
     def load_model(self) -> None:
         """Load model from PyTorchHub"""
         print("Loading YOLOv5 Torch model ...")
         if os.path.exists(self.model_local_path):
```
Comment (FatmaYoussef96) [inline src/dronevis/detection_torch/yolov5_torch.py:33]: since model_local_path is a class attribute, I think you should reference it using the class name instead of self

File: src/dronevis/detection_torch/yolov5_torch.py [hunk 5]
```diff
@@ -15,18 +17,27 @@ class YOLOv5(CVModel):
...
-            self.net = torch.hub.load(self.model_local_path, "yolov5s", source="local")
+            self.net = torch.hub.load(
```
Comment (FatmaYoussef96) [inline src/dronevis/detection_torch/yolov5_torch.py:34]: since model_local_path is a class attribute, I think you should reference it using the class name instead of self,
and the same goes for the other class variables

File: src/dronevis/drone_connect/command.py [hunk 6]
```diff
@@ -69,85 +91,81 @@ def configure(self, argument: str, value: str) -> bool:
             bool: flag for valid config operations
         """
         # Check if it's the first time we send a config
-        DEBUG = True
-        if not self.is_configurated:
+        if not self.is_configured:
```
Comment (FatmaYoussef96) [inline src/dronevis/drone_connect/command.py:94]: I think is_configured should be referenced using the class name instead of self as it is class variable

File: src/dronevis/drone_connect/nav_data_decode.py [hunk 7]
```diff
@@ -35,75 +37,83 @@ def _drone_status_decode(packet):
...
     offset = 0
-    block=[]
-    block.append(struct.unpack_from("=IIII",packet,position))
+    block = []
+    block.append(struct.unpack_from("=IIII", packet, position))
     offset += struct.calcsize("=IIII")
-    i=1
+    i = 1
 
     while 1:
```
Comment (FatmaYoussef96) [inline src/dronevis/drone_connect/nav_data_decode.py:99]: I think it should be while True to be consistent with other files in the project.

File: src/dronevis/drone_connect/video.py [hunk 8]
```diff
@@ -37,24 +42,26 @@ def run(self) -> None:
...
             prev_time = cur_time
-            cv2.imshow(self.frame_name, image)
+            cv2.imshow(self.frame_name, write_fps(frame, fps))
+
             if cv2.waitKey(1) & 0xFF == ord("q"):
                 break
-            
-        print("Closing video stream")
+
+        print("Closing video stream ...")
```
Comment (FatmaYoussef96) [inline src/dronevis/drone_connect/video.py:61]: I am not sure if this statement should be in logging or not.

**Comment by FatmaYoussef96** [comment]
Overall, great work. Thanks for your effort.
I left some comments on the changed files please read them and change the files if needed.
I have also some notes regarding drone-gui kindly take them into consideration:
- yolo took a lot of time without opening the webcam stream however it runs just in the detection with yolo example.
- same issue with faster R-CNN.

**Review by FatmaYoussef96 — COMMENTED** [review]
Overall, great work. Thanks for your effort.
I left some comments on the changed files please read them and change the files if needed.
I have also some notes regarding drone-gui kindly take them into consideration:
- yolo took a lot of time without opening the webcam stream however it runs just in the detection with yolo example.
- same issue with faster R-CNN.

File: README.md (hunk 2 above)
Comment (ahmedheakl) [reply README.md:42]: Maybe we should discuss a better alternative to represent the features in the next meeting.

File: src/dronevis/detection_torch/yolov5_torch.py (hunk 4 above)
Comment (ahmedheakl) [reply src/dronevis/detection_torch/yolov5_torch.py:33]: It wouldn't matter, as assigning a variable as a class attribute is just for code integrity, i.e., this variable should be persisted throughout all created instances from the same class. Moreover, in the class implementation it makes more sense to use `self` rather than the class name to ensure we are calling the class that we are currently developing. If it was on a different module other than the class's, your point would be valid.

**Comment by ahmedheakl** [comment]
Thank you for the review, will modify and mention you for another review later.

**Review by FatmaYoussef96 — COMMENTED** [review]
Great Job.
Just a gentle reminder of some issues when running the GUI on Windows platform to be addressed in future PRs:
1- segmentation of pose feature: throws an error " got an unexpected keyword: enable_segmentation is not defined".
2- Yolo and face models take a lot of time and don't open the webcam for detection.
